File: kata/bowling_game/go/2023_12-31/bowling_game.go
package bowling

type Game struct {
	score       int
	rolls       [21]int
	currentRoll int
}

func NewGame() *Game {
	return &Game{
		score:       0,
		currentRoll: 0,
	}
}

func (g *Game) Roll(pins int) {
	g.rolls[g.currentRoll] = pins
	g.currentRoll++
}

func (g *Game) Score() int {
	score := 0
	frameIndex := 0

	for frame := 0; frame < 10; frame++ {
		if g.isStrike(frameIndex) {
			score += g.strikeBonus(frameIndex)
			frameIndex++
		} else if g.isSpare(frameIndex) {
			score += g.spareBonus(frameIndex)
			frameIndex += 2
		} else {
			score += g.sumOfBallsInFrame(frameIndex)
			frameIndex += 2
		}
	}

	return score
}

func (g *Game) isStrike(frameIndex int) bool {
	return g.rolls[frameIndex] == 10
}

func (g *Game) isSpare(frameIndex int) bool {
	return g.rolls[frameIndex]+g.rolls[frameIndex+1] == 10
}

func (g *Game) sumOfBallsInFrame(frameIndex int) int {
	return g.rolls[frameIndex] + g.rolls[frameIndex+1]
}

func (g *Game) spareBonus(frameIndex int) int {
	return 10 + g.rolls[frameIndex+2]
}

func (g *Game) strikeBonus(frameIndex int) int {
	return 10 + g.rolls[frameIndex+1] + g.rolls[frameIndex+2]
}

File: kata/bowling_game/c/2023_12_10/src/bowling_game.c
// bowling_game.c
#include "bowling_game.h"

#include <stdlib.h>
#include <string.h>

void bowling_game_roll(BowlingGame* game, int pins) {
  game->current_score += pins;
  game->rolls[game->current_roll++] = pins;
}

int bowling_game_score(BowlingGame* game) {
  int score = 0;
  int frame_index = 0;
  for (int frame = 0; frame < 10; frame++) {
    if (is_strike(game, frame_index)) {
      score += 10 + strike_bonus(game, frame_index);
      frame_index++;
    } else if (is_spare(game, frame_index)) {
      score += 10 + spare_bonus(game, frame_index);
      frame_index += 2;
    } else {
      score += sum_of_balls_in_frame(game, frame_index);
      frame_index += 2;
    }
  }

  return score;
}

int sum_of_balls_in_frame(BowlingGame* game, int frame_index) {
  return game->rolls[frame_index] + game->rolls[frame_index + 1];
}

int spare_bonus(BowlingGame* game, int frame_index) {
  return game->rolls[frame_index + 2];
}

int strike_bonus(BowlingGame* game, int frame_index) {
  return game->rolls[frame_index + 1] + game->rolls[frame_index + 2];
}


int is_strike(BowlingGame* game, int frame_index) {
  return game->rolls[frame_index] == 10;
}

int is_spare(BowlingGame* game, int frame_index) {
  return game->rolls[frame_index] + game->rolls[frame_index + 1] == 10;
}

BowlingGame* new_bowling_game() {
  BowlingGame* game = malloc(sizeof(BowlingGame));
  game->current_roll = 0;
  memset(game->rolls, 0, sizeof(game->rolls));
  game->current_score = 0;
  game->roll = bowling_game_roll;
  game->score = bowling_game_score;
  return game;
}
